src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs

```diff
@@ -127,6 +127,11 @@ public void Advance(int count)
             Debug.Assert(_index <= _rentedBuffer.Length - count);
 
             _index += count;
+
+            if ((uint)_index > int.MaxValue)
```

Comment:
could this be simply `_index < 0`?

Reply:
also you might wanna throw before the Advance, I'm not quite sure if this broken internal state can be re-used anywhere

Reply:
Based on the assert on line 127, I don't think this check could ever be true?

Reply:
Furthermore, I don't think this is the right method to incorporate such a check. It should live in `CheckAndResizeBuffer`. It seems that the method already has a check to that effect:

https://github.com/dotnet/runtime/blob/ebdb045532190ffc664bba9a0a1e3f2ce35cf23f/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs#L187-L194

But it might be that it's too conservative? We might want to replace `int.MaxValue` with `Array.MaxLength`.

Reply:
Per https://github.com/dotnet/runtime/issues/73992#issuecomment-1238447939 I think the logic here is sound. `Array.MaxLength` isn't available on netstandard2.0 and net462 but we could use it where available.